pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+			// this function, we'd have successfully deleted the old rule which would steer
+			// traffic towards these routes.
+			if err := m.deleteOldRoutes(routes); err != nil {
+				scopedLog.WithError(err).WithField("routes", routes).
+					Warn("Failed to cleanup after successfully migrating endpoint to original ENI datapath. " +
+						"It is recommended that the old rule and routes are cleaned up, as it is possible in the future " +
+						"to collide without another endpoint with the same IP.")
+			}
+
+			migrated++
+		}
+	}
+
+	return migrated, failed, partial
+}
+
+// NewMigrator constructs a migrator object with the default implementation to
+// use the underlying upstream netlink library to manipulate the Linux RPDB.
+// It accepts a getter for retrieving the interface number by MAC address and
+// vice versa.
+func NewMigrator(getter interfaceDB) *migrator {
+	return &migrator{
+		rpdb:   defaultRPDB{},
+		getter: getter,
+	}
+}
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. The table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex, plus the specific table offset
+// value.
+func (m *migrator) retrieveTableIDFromIfIndex(ifindex int) (int, error) {
+	link, err := m.rpdb.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := m.getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	// This is guaranteed to avoid conflicting with the main routing table ID
+	// (253-255) because the maximum number of ENI devices on a node is 15 (see
+	// pkg/aws/eni/limits.go). Because the interface number is monotonically
+	// increasing and the lowest available number is reused when devices are
+	// added / removed. This means that the max possible table ID is 25.
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// retrieveTableIDFromInterfaceNumber returns the table ID based on the
+// interface number. The table ID is the ifindex of the device corresponding to
+// the ENI with the given interface number. This is used for downgrading /
+// using the old ENI datapath.
+func (m *migrator) retrieveTableIDFromInterfaceNumber(ifaceNum int) (int, error) {
+	mac, err := m.getter.GetMACByInterfaceNumber(ifaceNum)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	links, err := m.rpdb.LinkList()
+	if err != nil {
+		return -1, fmt.Errorf("failed to list links: %w", err)
+	}
+
+	var link netlink.Link
+	for _, l := range links {
+		if l.Attrs().HardwareAddr.String() == mac {
+			link = l
+			break
+		}
+	}
+
+	return link.Attrs().Index, nil
+}
+
+func (m *migrator) copyRoutes(routes []netlink.Route, from, to int) (revert.RevertStack, error) {
```

Comment:
There's a particular assumption that this function makes which is not called out, and is subtly different from some other code:

This function could return three possible results:
* No reverstack and an error
* Revertstack (to revert first successful operation) and an error
* Reverstack (to revert all operations) and no error

Therefore, the caller *must* run `revertstack.Revert()` while handling the error and if they miss that, it's a bug. I note that the code today seems to do this pretty consistently so I think it's correct. That said, this is an assumption the code makes.

The caller can be simplified slightly by streamlining the possible return cases to:
* No reverstack and an error
* Reverstack (to revert all operations) and no error

There's multiple ways to handle this, one would be to add a defer which checks the error result & automatically runs the revertstack if an error occurs. Another common way in other languages is using a goto for a dedicated error handling section at the end of the function.

I don't think we necessarily have to switch to the latter approach, but ideally we should call out this assumption above the function signature so that the expectations for the caller are more clear. (Conversely, if we *do* change this, then all callers must be checked for their usage of the reverstack to conform to whichever model this function assumes)

Reply:
Updated the godoc on `copyRoutes` and `createNewRule`.